### PR TITLE
Development: Push neural net packages and NumPy into a separate optional dependency group to reduce wait on cold setup.

### LINKED ISSUE #278: [Development] Exclude `torch` and its dependencies into a separate group of optional dependencies.
Whenever I've created/updated a virtual environment before making a PR for Tyro, I've had to spend a number of minutes waiting for PyTorch and its dependencies (particularly the CUDA packages) to download. While not a huge deal, it is a bit of a barrier. Maybe a separate `dev-torch` group under `optional-dependencies` would be worthwhile, along with a way to exclude certain tests, which depend on it, from running? Alternatively, maybe there is a way to mock the PyTorch dependencies to remove this ultra-heavy dependency tree entirely?

Again, not a big deal... just throwing out a suggestion.

### PR DESCRIPTION
**Pull request by emcd**
Closes #278 .

Passes Ruff checks in all environments.
Passes Pyright analysis in `dev-nn` environments < Python 3.13.
Test suite passes in all environments.

Sizes:
```text
$ du -hs venv-3.10/lib/python3.10/site-packages
139M    venv-3.10/lib/python3.10/site-packages
$ du -hs venv-3.13/lib/python3.13/site-packages
137M    venv-3.13/lib/python3.13/site-packages
$ du -hs venv-nn-3.10/lib/python3.10/site-packages
6.0G    venv-nn-3.10/lib/python3.10/site-packages
$ du -hs venv-nn-3.13/lib/python3.13/site-packages
204M    venv-nn-3.13/lib/python3.13/site-packages
```

A bit messier than I had hoped. Imports being used in module-level class definitions prevented use of `pytest.mark.skipif` as I had hoped. And using it where it could be used (in conjunction with `try .. except ImportError`) made Pyright upset since static analysis obviously could not understand the semantics of `skipif`.

Wasn't feeling motivated to work on any of my projects this evening, which is why I did this, even though the effort-reward ratio is rather questionable. Obviously not a high priority for review.

### CODE REVIEW

**Review by brentyi — APPROVED** [review]
Looks reasonable to me. Thanks @emcd!
